### ISSUE TITLE
Type is not infered from default value of dict.get()

### ISSUE THREAD

**Issue by qi55wyqu** [python/mypy#10890]
**Bug Report**

When using `dict.get(value, default)` the returned type is not inferred correctly if the value was initially declared as `Optional`.

**To Reproduce**

```python
mapping: Dict[str, str] = {"a": "x"}
val: Optional[str] = None
val = "a"
reveal_type(val) # Revealed type is "builtins.str"
val = mapping[val] if val in mapping else val
reveal_type(val) # Revealed type is "builtins.str"
val = mapping.get(val, val)
reveal_type(val) # Revealed type is "Union[builtins.str, None]"
```

**Expected Behavior**

Revealed type is `builtins.str`.

**Actual Behavior**

Revealed type is `Union[builtins.str, None]`.

**Your Environment**

- Mypy version used: mypy 0.910
- Mypy command-line flags: --
- Mypy configuration options from `mypy.ini` (and other config files): --
- Python version used: 3.8.10
- Operating system and version: Ubuntu 20.04.2

**Comment by JelleZijlstra** [comment]
Isn't this because `val` is of type `Optional[str]`?

**Comment by erictraut** [comment]
The declared type of `val` is `Optional[str]`, but at the time it is used in the expression `mapping.get(val, val)`, its type has been narrowed to just `str`, so I would expect the return type of this call expression to be `str` in this case. That's the type pyright computes.

**Comment by hauntsaninja** [comment]
This is another "type context" thing. If you use a new variable, you get the result you want, i.e.:
```
new_val = mapping.get(val, val)
reveal_type(new_val)
```

**Comment by qi55wyqu** [comment]
```python
mapping[v] if v in mapping else v
```
and 
```python
mapping.get(v, v)
```
are basically the same thing so I would expect mypy to infer the same type for the return value in both cases.

**Comment by AlexWaygood** [comment]
Fixed in #14151
